### PR TITLE
Ldelf nospec

### PR DESCRIPTION
**Pull request by jenswi-linaro**
"ldelf: use confine_array_index() to cap speculation" brings protection against Spectre V1 speculations.

When parsing the ELF there's indexes and offsets used to load data which in turn is used to load other data and so on, a growing ground for Spectre V1. I think `confine_array_index()` should only be used when needed or sometimes needed. It is however not so easy to tell since it's with the second data load there's danger. This ELF parsing is a bit of a special case so perhaps we should try to annotate the code in a way to assist someone reviewing the code to see if another call to `confine_array_index()` might be needed or not.

### CODE REVIEW

File: ldelf/ta_elf_rel.c [hunk 1]
```diff
@@ -185,6 +202,12 @@ static void e32_relocate(struct ta_elf *elf, unsigned int rel_sidx)
 
 		str_tab_idx = shdr[sym_tab_idx].sh_link;
 		if (str_tab_idx) {
+			if (str_tab_idx >= elf->e_shnum)
+				err(TEE_ERROR_GENERIC,
```
Comment (jforissier) [inline ldelf/ta_elf_rel.c]: `TEE_ERROR_BAD_FORMAT` would probably be better, but I see `TEE_ERROR_GENERIC` is used in similar places, so if we want to change it it should be in a separate patch.

File: ldelf/ta_elf_rel.c [hunk 2]
```diff
@@ -185,6 +202,12 @@ static void e32_relocate(struct ta_elf *elf, unsigned int rel_sidx)
...
+				    "Strtab index out of range");
```
Comment (jforissier) [inline ldelf/ta_elf_rel.c]: `STRTAB` (all uppercase) perhaps? That's the way `readelf` prints it.

File: ldelf/ta_elf_rel.c [hunk 3]
```diff
@@ -317,6 +342,12 @@ static void e64_relocate(struct ta_elf *elf, unsigned int rel_sidx)
...
```
Comment (jforissier) [inline ldelf/ta_elf_rel.c]: Same

File: ldelf/ta_elf_rel.c [hunk 4]
```diff
@@ -42,11 +42,14 @@ static bool __resolve_sym(struct ta_elf *elf, unsigned int bind,
...
 		return false;
 	if (st_name > elf->dynstr_size)
-		err(TEE_ERROR_BAD_FORMAT, "Symbol out of range");
+		err(TEE_ERROR_BAD_FORMAT, "Symbol name out of range");
 
 	if (strcmp(name, elf->dynstr + st_name))
 		return false;
 
+	if (st_value > (elf->max_addr - elf->load_addr))
+		err(TEE_ERROR_BAD_FORMAT, "Symbol location out of range");
```
Comment (jforissier) [inline ldelf/ta_elf_rel.c]: Hmmmm... not always OK. Depends on the symbol type. For "usual" symbols that we may encounter at load time (`STT_OBJECT` and `STT_FUNC` for instance), we do expect an address. But, consider the `STT_TLS` type for instance. The [spec](https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-79797.html) says:

> STT_TLS
> The symbol specifies a thread-local storage entity. When defined, this symbol gives the assigned offset for the symbol, not the actual address.

I happen to know this well, because I'm currently working on C++ support and in order to be able to use the `arm-linux-gnueabihf-` toolchain, I had to implement a relocation type which will typically lookup such symbols.

That being said, the fact that we're adding `elf->load_addr` assumes that this function can only be used for symbols that represent addresses...

How about passing the symbol type to the function? Then the proper check can be done and `elf->load_addr` can be added only when applicable. Which will integrate well with my future patches :wink:

File: ldelf/ta_elf_rel.c (hunk 1 above)
Comment (jenswi-linaro) [reply ldelf/ta_elf_rel.c]: Sounds good. I'll follow up this PR with another to make error codes more consistent.

File: ldelf/ta_elf_rel.c (hunk 2 above)
Comment (jenswi-linaro) [reply ldelf/ta_elf_rel.c]: Makes sense, but there's quite a few other places which should be updated too then.
OK if I take that in a follow up PR so I can update all the other places too?

Comment (jforissier) [reply ldelf/ta_elf_rel.c]: Fine by me

File: ldelf/ta_elf_rel.c (hunk 4 above)
Comment (jenswi-linaro) [reply ldelf/ta_elf_rel.c]: Makes sense, I'll add and the check symbol type.

File: ldelf/ta_elf.c [hunk 5]
```diff
@@ -733,6 +736,7 @@ static void populate_segments(struct ta_elf *elf)
 				if (res)
 					err(res, "sys_copy_from_ta_bin");
 			} else {
+				assert(filesz == memsz);
```
Comment (etienne-lms) [inline ldelf/ta_elf.c]: assert here or return an error?

Comment (jenswi-linaro) [reply ldelf/ta_elf.c]: I added this for debug purpose, but I don't see any harm in returning error instead. I'll update.

**Comment by jenswi-linaro** [comment]
update

**Comment by jenswi-linaro** [comment]
Squashed and tags applied

**Comment by jforissier** [comment]
Can you please fix the spelling mistake in the last patch?

**Comment by jenswi-linaro** [comment]
Update
